lib/docs/query.rb

```diff
@@ -27,7 +27,8 @@ def list
 
       # filter results by selected category
       if @filters[:category].present?
-        results = results.where('topics.category_id IN (?)', @filters[:category].split('|'))
+        category_ids = @filters[:category].split('|')
+        results = results.where('topics.category_id IN (?)', category_ids) if category_ids.all? { |id| id =~ /\A\d+\z/ }
```

Comment:
Hmm I'm not sure if this is sufficient. Do we have to check that the user has access to all the categories before passing the ids through to the query?

Reply:
Good point, I ensured we are safe. I tried to access private category as regular user, and it is not possible.

At the top we have
```
results = tq.list_docs_topics
```
```
add_to_class(:topic_query, :list_docs_topics) do
    default_results(@options)
  end
```

and default results already have check:
https://github.com/discourse/discourse/blob/main/lib/topic_query.rb#L696

Reply:
I think we need to play it safer here because we do not know what might change in `TopicQuery#default_results` in the future so we cannot assume that the method will always handle checking categories for us. I think we should just check the ids against the guardian here.

Another thing that is strange to me is the filter above `results = results.where('topics.category_id IN (?)', Query.categories).or(results.where('tags.name IN (?)', Query.tags))`. I think if `@filters[:category]` is present we should be doing an intersection with `Query.categories`.

Sorry not related to your changes so I'm happy to approve as it is now.